[Website] Menu and Menu buttons docs (#3765)
carbon-design-system/carbon-website: src/pages/components/menu-buttons/usage.mdx

| [Menu button](#menu-button)     | Use a menu button when all actions in the menu list share the same level of importance. Usually used in the page header.                                     |
| [Combo button](#combo-button)   | Use a combo button when screen real estate is limited and one of the actions has particular importance. They are usually used in the page header.            |
| [Overflow menu](#overflow-menu) | Use an overflow button when additional options are available to the user and there is a space constraint. They are usually used in tiles or data table rows. |

## Live demo

<StorybookDemo
  themeSelector
  url="https://react.carbondesignsystem.com"
  short
  variants={[
    {
      label: 'Menu button',
      variant: 'components-menubutton--default',
    },
    {
      label: 'Combo button',
      variant: 'components-combobutton--default',
    },
    {
      label: 'Overflow menu',
      variant: 'components-overflowmenu--default',
    },
  ]}
/>

## Formatting

### Anatomy

<Row>
<Column colLg={8}>

> ![Menu button anatomy diagram](images/usage-anatomy.png)

Reviewer: This is still an issue

[fluid-width default buttons](https://carbondesignsystem.com/components/button/usage/#alignment),
both the menu button (or the combo button) and the menu should have the same
width.

<DoDontRow>
  <DoDont type="do" caption="Do use the minimum width">

![Do use the minimum width of menu](images/usage-menu-width-do-1.png)

  </DoDont>
  <DoDont type="dont" caption="Do not narrow the menu to fit with other elements">

![Do not narrow the menu to fit with other elements](images/usage-menu-width-dont-1.png)

  </DoDont>
</DoDontRow>

<DoDontRow>
  <DoDont type="do" caption="Do match the width when the menu or combo button label becomes longer">

![Do match the width when the menu or combo button labels become longer](images/usage-menu-width-do-2.png)

  </DoDont>
  <DoDont type="dont" caption="Do not make the menu appear shorter than the menu or combo buttons">

![Do not make the menu appear shorter than the menu or combo buttons](images/usage-menu-width-dont-2.png)

  </DoDont>
</DoDontRow>

Ghost buttons only adhere to the
[button](https://carbondesignsystem.com/components/button/style/) style. This is
because enlarging the ghost button's width to match the menu's width would

> result in the chevron appearing disconnected from its associated button,

Reviewer: Oh this was my bad, a chevron is correct here!

available space and layout.

<Row>
<Column colLg={8}>

![Menu default alignments](images/usage-menu-alignment-menu-buttons-default.png)

</Column>
</Row>

#### Alternatives

Alternatively, open menus can be positioned at the bottom, to the left or right
of its associated menu button trigger, depending on the available space and
layout.

<Row>
<Column colLg={8}>

![Menu alternative alignments](images/usage-menu-alignment-menu-buttons-alternatives.png)

</Column>
</Row>

### Placement

#### Menu and combo buttons

These buttons are ideal for page headers, where multiple, distinct, actions may
impact the page content, and where real-estate is limited. In the page header
component, multiple buttons will become a single action button at smaller screen
sizes. These buttons should only be used when the actions affect a full page or
a large object on a page (examples include a canvas UI, diagram or whole data
table).

<Row>
<Column colLg={8}>

![Menu button alignment example](images/usage-menu-placement-1.png)

</Column>
</Row>

#### Overflow menu

When actions affect a smaller proportion of a page, such as a table rows or
cards, use an overflow menu.

<Row>
<Column colLg={8}>

> ![Overflow menu alignment example](images/usage-menu-placement-2.png)

Reviewer: The overflow menu in this image has a defect with the shadow coming on top where the button and menu meet. The shadow should be behind the button and menu.

</Column>
</Row>

## Content

### Main elements

#### Label

The text label of the button is important to communicate the action that is
triggered when the user interacts with it. The text label must accurately
describe the button action or indicate that interacting with it displays a list
of possible actions.

### Further guidance

For further content guidance, see Carbon's
[content guidelines](/guidelines/content/overview).

## Universal behaviors

### States

> The default and interactive states of menu buttons should follow the button

Reviewer: Did you mean to link out to the button style guidelines here?

style guidelines.

- Combo buttons are available only as primary buttons.
- Menu buttons come in three different button types: primary, tertiary, and
  ghost.

> - Overflow button is treated as the ghost button, displaying only an icon and

Reviewer: ```suggestion
- Overflow menu is treated as a ghost button, displaying only an icon and
```

Author: hmm I was thinking about this..wondering if it should be called icon or menu...because it's a button, I'll try "the icon button of overflow menu"

  following its designated color and style.

### Interactions

#### Menu button

Users can reveal the menu by:

- Clicking anywhere on the field of a menu button
- Or tabbing into the menu button and pressing `Return` or `Enter`

<Row>
<Column colLg={8}>

![Menu button interactions](images/usage-menu-keyboard-menu-button.png)

<Caption>Mouse and keyboard interactions for menu button</Caption>
</Column>
</Row>

#### Combo button

Users can reveal the menu by:

- Clicking anywhere on the field of the primary button or the icon button. If
  clicking on the primary button, this will apply to certain elements, depending
  on user interaction.
- Or tabbing into the combo button which will focus on the primary button first
  then the icon button. Then, pressing the `Return` or `Enter` to trigger the

>   action of the primary button or menu under icon button.

Reviewer: ```suggestion
  action of the primary button or menu under the icon button.
```

</Column>
</Row>

<Row>
<Column colLg={8}>

![Combo button keyboard interaction](images/usage-menu-keyboard-combo-button.png)

<Caption>Keyboard interactions for combo button</Caption>
</Column>
</Row>

#### Overflow menu

Users can reveal the overflow menu by:

- Clicking anywhere on the field of the icon button
- Or tabbing into the icon button and pressing `Return` or `Enter`

<Row>
<Column colLg={8}>

![Overflow menu mouse and keyboard interactions](images/usage-menu-mouse-keyboard-overflow-menu.png)

<Caption>Keyboard interactions for overflow menu</Caption>
</Column>
</Row>

## Menu button

Menu buttons group actions that share the same level of importance. The icon
used within the menu button should always be a chevron in order to visually
convey to the user that there are options to select from. There are three main
use cases where using a menu button is recommended:

- Grouping objects with the same common action
- Actions related to a common title
- Distinct actions applied to the same object or page

Menu button and combo button come in three different button types: primary,
tertiary, and ghost.

<Row>
<Column colLg={8}>

> ![An example of menu button](images/usage-menu-button-1.png)

Reviewer: The menu should be using `$layer-02` background here since its opening on top of $layer-01.

</Column>
</Row>

### Best practices

#### Grouping objects with the same common action

Menu buttons are effective when a single action affects multiple distinct
objects. This is particularly applicable when users need to Create or Add
various objects, and all these objects are equally relevant options.

The shared action should not be repeated in the menu list. This adds unnecessary
clutter and makes each option less distinct at a glance.

<DoDontRow>
  <DoDont type="do" caption="Do add distinct items for the action">

![Do add distinct items for the action](images/usage-menu-button-2.png)

</DoDont>
<DoDont type="dont" caption="Do not repeat the action in the action item">

![Do not repeat the action in the action item](images/usage-menu-button-3.png)

  </DoDont>
</DoDontRow>

#### Actions associated with a shared label

Another common use of a menu button is grouping actions related to a common

> title. An example of this is when users have multiple approaches to execute an

Reviewer: Should this be "label", not title?

action, all of which are presented as choices within the menu button.

<Row>
<Column colLg={8}>

> ![Actions associated with a shared label](images/usage-menu-button-4.png)

Reviewer: The chevron icon in the button should be pointed up in this image

</Column>
</Row>

#### Distinct actions applied to the same object or page

In many cases, actions on a page may not use the same verb or have a shared
label for the menu button. When grouped actions remain distinct, consider
labeling the menu button as 'Actions' to encompass all available choices.

<Row>
<Column colLg={12}>

> ![Distinct actions applied to the same object or page](images/usage-menu-button-5.png)

Reviewer: Also looks like the text in the green tags in the data table are magenta/purple? They should be the darker green color for tags.

> ![Distinct actions applied to the same object or page](images/usage-menu-button-5.png)

Author: I think it's a bug for the dark green color.... oh boy one of those bug issues..

primary button so it is not hidden within the menu of additional actions. The
primary default action is typically the most commonly used action there is to
take.

<Row>
<Column colLg={8}>

![Choose a default action](images/usage-menu-combo-button-2.png)

<Caption>
  Example shows that users mostly choose to save records, while less common
  options are hidden as extra actions.
</Caption>
</Column>
</Row>

#### Reduce visual complexity on a page

Combo buttons reduce visual complexity by grouping similar commands together.
For example, how navigation menus group together related options to enable
conceptual understanding of the site information structure.

<Row>
<Column colLg={8}>

![Reduce visual complexity on a page](images/usage-menu-combo-button-3.png)

<Caption>
  Example shows that many actions (ghost buttons) applying to the same data can
  be replaced by the combo button.
</Caption>
</Column>
</Row>

## Overflow menu

Overflow menu is also another option that is available to the user when there is
a space constraint, usually used in data table rows or in tiles. Overflow menu
has the ghost button style only.

<Row>
<Column colLg={8}>

> ![An example of overflow menu](images/usage-overflow-1.png)

Reviewer: - The overflow menu in this image has a defect with the shadow coming on top where the button and menu meet. The shadow should be behind the button and menu.
- The search icon color should be the same as the other toolbar icon colors. Looks like it might be at $icon-secondary currently.
- The status icons in the data table should have the Accessibility inner fill color.

</Column>
</Row>

### Best practices

#### Overflow menu in cards

When there are more than three interactive icons, consider moving less important
ones into the overflow menu. The overflow menu then takes the place of a third
action.

<Row>
<Column colLg={8}>

![An example of overflow menu in cards](images/usage-overflow-2.png)

</Column>
</Row>

#### Overflow menu icon

> Overflow icon can be customized to align with the product's requirements.

Reviewer: ```suggestion
The overflow icon can be customized to align with the product's requirements.
```

However, it's advisable to make minimal changes to the overflow button's icons
to maintain consistency across the product and ensure a consistent user
experience.

<Row>
<Column colLg={4}>

![An example of another overflow menu icon](images/usage-overflow-3.png)

</Column>
</Row>

#### Overflow versus tab tip (popover)

Overflow menu displays extra options with simple text, typically utilized within

> data table rows, tiles, or cards.

Reviewer: ```suggestion
data table rows or cards.
```

> [Tab tip](https://carbondesignsystem.com/components/popover/usage/#tab-tip)

Reviewer: ```suggestion
The [tab tip](https://carbondesignsystem.com/components/popover/usage/#tab-tip)
```

> popover provides more complex inputs such as a checkbox group or a radio button

Reviewer: ```suggestion
popover allows the user to include more complex inputs such as a checkbox group or a radio button
```

group, usually connected to a data table toolbar.

<Row>
<Column colLg={4}>

> ![An example of overflow menu](images/usage-overflow-menu.png)

Reviewer: The bottom border line of the left image does not extend all the way to the right edge of the illustration.

<Caption>Use overflow menus in data table row</Caption>
</Column>

<Column colLg={4}>

![An example of a popover](images/usage-popover.png)

<Caption>Use a popover in the data table toolbar</Caption>
</Column>
</Row>

## Related

#### Menu

To learn more about the details of how to use the menu component correctly, see

> [menu buttons](/components/menu-buttons/usage)

Reviewer: ```suggestion
[menu buttons](/components/menu-buttons/usage).
```

triggered when the user interacts with it. The text label must accurately
describe the button action or indicate that interacting with it displays a list
of possible actions.

### Further guidance

For further content guidance, see Carbon's
[content guidelines](/guidelines/content/overview).

## Universal behaviors

### States

The default and interactive states of menu buttons should follow the button
[style](https://carbondesignsystem.com/components/button/style) guidelines.

- Combo buttons are available only as primary buttons.
- Menu buttons come in three different button types: primary, tertiary, and
  ghost.
- The icon button of overflow menu is treated as the ghost button, displaying
  only an icon and following its designated color and style.

### Interactions

#### Menu button

Users can reveal the menu by:

- Clicking anywhere on the field of a menu button
- Or tabbing into the menu button and pressing `Return` or `Enter`

<Row>
<Column colLg={8}>

![Menu button interactions](images/usage-menu-keyboard-menu-button.png)

<Caption>Mouse and keyboard interactions for menu button</Caption>
</Column>
</Row>

#### Combo button

Users can reveal the menu by:

> - Clicking anywhere on the field of the primary button or the icon button. If

Reviewer: ```suggestion
- Clicking anywhere on the container of the primary button or the icon button. If
```

label for the menu button. When grouped actions remain distinct, consider
labeling the menu button as 'Actions' to encompass all available choices.

<Row>
<Column colLg={12}>

![Distinct actions applied to the same object or page](images/usage-menu-button-5.png)

<Caption>Example shows the menu button with possible actions that could influence the data within tab contents</Caption>
</Column>
</Row>

## Combo button

A combo button is composed of two buttons: the left is the default option (the
most commonly used selection), and the right (chevron icon) shows the rest of
the available options. A combo button provides quick access to a default action
while also offering a list of related actions or options.

The combo button is expected to have less use than the menu button, as the
additional actions are more hidden than they would be in menu button. However,
this button does have particular use where screen real estate is limited and one
of the actions has particular importance.

Combo button has only one button type: primary.

<Row>
<Column colLg={8}>

![An example of combo button](images/usage-menu-combo-button-1.png)

</Column>
</Row>

### Best practices

> #### Chose a default action

Reviewer: ```suggestion
#### Choose a default action
```

<Row>
<Column colLg={8}>

![Reduce visual complexity on a page](images/usage-menu-combo-button-3.png)

<Caption>
  Example shows that many actions (ghost buttons) applying to the same data can
  be replaced by the combo button.
</Caption>
</Column>
</Row>

## Overflow menu

Overflow menu is also another option that is available to the user when there is
a space constraint, usually used in data table rows or in cards. Overflow menu
has the ghost button style only.

<Row>
<Column colLg={8}>

![An example of overflow menu](images/usage-overflow-1.png)

</Column>
</Row>

### Best practices

#### Overflow menu in cards

When there are more than three interactive icons, consider moving less important
ones into the overflow menu. The overflow menu then takes the place of a third
action.

<Row>
<Column colLg={8}>

![An example of overflow menu in cards](images/usage-overflow-2.png)

</Column>
</Row>

#### Overflow menu icon

The overflow icon can be customized to align with the product's requirements.
However, it's advisable to make minimal changes to the overflow menu's icon to
maintain consistency across the product and ensure a consistent user experience.

<Row>

> <Column colLg={4}>

Reviewer: ```suggestion
<Column colLg={8}>
```

![An example of another overflow menu icon](images/usage-overflow-3.png)

</Column>
</Row>

#### Overflow versus tab tip (popover)

Overflow menu displays extra options with simple text, typically utilized within
data table rows or cards. The
[tab tip](https://carbondesignsystem.com/components/popover/usage/#tab-tip)

> popover allows the user to includes more complex inputs such as a checkbox group

Reviewer: ```suggestion
popover allows the user to include more complex inputs such as a checkbox group
```
